python-package/lightgbm/sklearn.py

```diff
@@ -662,17 +670,57 @@ def __init__(
         self._n_classes: int = -1
         self.set_params(**kwargs)
 
+    # scikit-learn 1.6 introduced an __sklearn__tags() method intended to replace _more_tags().
+    # _more_tags() can be removed whenever lightgbm's minimum supported scikit-learn version
+    # is >=1.6.
+    # ref: https://github.com/microsoft/LightGBM/pull/6651
     def _more_tags(self) -> Dict[str, Any]:
         return {
             "allow_nan": True,
             "X_types": ["2darray", "sparse", "1dlabels"],
             "_xfail_checks": {
                 "check_no_attributes_set_in_init": "scikit-learn incorrectly asserts that private attributes "
                 "cannot be set in __init__: "
-                "(see https://github.com/microsoft/LightGBM/issues/2628)"
+                "(see https://github.com/microsoft/LightGBM/issues/2628)",
+                "check_n_features_in_after_fitting": (
+                    "validate_data() was first added in scikit-learn 1.6 and lightgbm"
+                    "supports much older versions than that"
+                ),
             },
         }
 
+    @staticmethod
+    def _update_sklearn_tags_from_dict(
+        *,
+        tags: "_sklearn_Tags",
+        tags_dict: Dict[str, Any],
+    ) -> "_sklearn_Tags":
+        """Update ``sklearn.utils.Tags`` inherited from ``scikit-learn`` base classes.
+
+        ``scikit-learn`` 1.6 introduced a dataclass-based interface for estimator tags.
+        ref: https://github.com/scikit-learn/scikit-learn/pull/29677
+
+        This method handles updating that instance based on the value in ``self._more_tags()``.
+        """
+        tags.input_tags.allow_nan = tags_dict["allow_nan"]
+        tags.input_tags.sparse = "sparse" in tags_dict["X_types"]
+        tags.target_tags.one_d_labels = "1dlabels" in tags_dict["X_types"]
+        tags._xfail_checks = tags_dict["_xfail_checks"]
+        return tags
+
+    def __sklearn_tags__(self) -> Optional["_sklearn_Tags"]:
+        # _LGBMModelBase.__sklearn_tags__() cannot be called unconditionally,
+        # because that method isn't defined for scikit-learn<1.6
+        if not callable(getattr(_LGBMModelBase, "__sklearn_tags__", None)):
+            return None
```

Comment:
I would personally prefer using an `available_if` here since now this logic is not less complicated as having that one. But this works too. However, maybe raising an `AttributeError` would be better? This method doesn't need to exist in older sklearn versions.

Reply:
Do you mean `sklearn.utils.available_if`?

https://github.com/scikit-learn/scikit-learn/blob/74a33757c8a8df84d227f28bbc9ec7ae2fb51dea/sklearn/utils/_available_if.py#L57

I prefer this method with `getattr()` that doesn't take on another `sklearn` import that could possibly be moved or changed in future versions.

Reply:
Fair enough, but `available_if` is kinda part of our _public_ developer API, so we'd think twice before changing / moving it. But it's not impossible to change.

Reply:
> *However, maybe raising an `AttributeError` would be better?*

I realize I was only focused on the `available_if()` part of your suggestion and didn't think carefully about the `AttributeError` part. I like that! Having fewer `Optional[]` or `{type} | None` type hints is really helpful... those possibilities of `None` tend to permeate through other calling code.

Reply:
I've implemented this in recent commits, and added a test around it.